### PR TITLE
Add SymFloat, support SymInt to SymFloat conversion

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #84284

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by wconstab** [comment]
can you add the motivation for this and also the impact to LTC/XLA?

I guess it's to handle things like .sym_sizes(0) / 2.0 properly, and that makes sense, but it'd be good to have the assumptions captured.

File: c10/core/SymFloat.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+#include <memory>
+
+namespace c10 {
+
+// NB: this is actually double precision; we're using the Python naming here
+class C10_API SymFloat {
+ public:
+  /*implicit*/ SymFloat(double d) : data_(d){};
+  SymFloat(SymFloatNode ptr)
+      : data_(std::numeric_limits<double>::quiet_NaN()), ptr_(std::move(ptr)){};
```
Comment (wconstab) [inline c10/core/SymFloat.h:18]: hm, so this deviates from SymInt with its negative number trick out of necessity due to bit properties of the types, and the net effect for now is the type is twice as big?

Comment (ezyang) [reply c10/core/SymFloat.h:18]: I could NaN box it but there is no reason to right now, this is not perf critical.

**Comment by ezyang** [comment]
> I guess it's to handle things like .sym_sizes(0) / 2.0 properly, and that makes sense, but it'd be good to have the assumptions captured.

SymFloat does help with that, but it's actually not the point (and I didn't add enough support for that); the real reason it exists is so I can type promote SymInt into SymFloat and then do tensor / sym_float which shows up in normalization decomps.

File: docs/source/conf.py [hunk 2]
```diff
@@ -322,7 +322,8 @@
     "Quantize",
     # torch.utils.backcompat
     "Warning",
-    "SymIntNode"
+    "SymIntNode",
```
Comment (albanD) [inline docs/source/conf.py:325]: There should really either have a doc entry on the website or be private...

File: torch/_C/__init__.pyi.in [hunk 3]
```diff
@@ -171,6 +171,12 @@ class SymIntNode(object):
     @staticmethod
     def new_symint(self) -> SymIntNode: ...
 
+class SymFloatNode(object):
+    def get_pyobj(self) -> Any: ...
+
+    @staticmethod
+    def new_symfloat(self) -> SymFloatNode: ...
```
Comment (albanD) [inline torch/_C/__init__.pyi.in:178]: nit: static method so the arg is `obj` and not `self`

File: torch/csrc/jit/python/init.cpp [hunk 4]
```diff
@@ -158,6 +161,9 @@ class PythonSymIntNodeImpl : public c10::SymIntNodeImpl {
     return getPyObj().attr("__int__")().cast<int64_t>();
   }
 
+  // TODO: virtualize
+  SymFloat sym_float();
```
Comment (albanD) [inline torch/csrc/jit/python/init.cpp:165]: On my local build in the branch I had to add `c10::` here. But looks ok here.

File: torch/csrc/utils/python_arg_parser.cpp [hunk 5]
```diff
@@ -732,7 +732,11 @@ auto FunctionParameter::check(
       if (is_tensor_and_append_overloaded(obj, &overloaded_args)) {
         return true;
       }
-      return allow_numbers_as_tensors && THPUtils_checkScalar(obj);
+      if (allow_numbers_as_tensors) {
+        return THPUtils_checkScalar(obj) ||
+            torch::is_symint_node(py::handle(obj));
```
Comment (albanD) [inline torch/csrc/utils/python_arg_parser.cpp:737]: or symfloat?

File: torch/csrc/jit/python/pybind_utils.cpp [hunk 6]
```diff
@@ -63,12 +66,27 @@ IValue toIValue(py::handle obj, const TypePtr& type, c10::optional<int32_t> N) {
           scalar = at::Scalar(THPUtils_unpackComplexDouble(obj.ptr()));
         } else if (THPUtils_checkDouble(obj.ptr())) {
           scalar = at::Scalar(THPUtils_unpackDouble(obj.ptr()));
+        } else if (torch::is_symint_node(py::handle(obj))) {
+          save_symint = true;
+          scalar = at::Scalar(7777777);
```
Comment (albanD) [inline torch/csrc/jit/python/pybind_utils.cpp:71]: Do we want to #define this constant to ensure it remains consistent from all the files?

File: torch/fx/experimental/proxy_tensor.py [hunk 7]
```diff
@@ -454,6 +454,7 @@ def assert_proxy_tensor(e):
 
 
 SymInt = torch._C.SymIntNode
+SymFloat = torch._C.SymFloatNode
```
Comment (albanD) [inline torch/fx/experimental/proxy_tensor.py:457]: We dump these in the torch. namespace since they don't look private, so you can access them directly on `torch.SymFloatNode` !
See other comment about making these private or documenting them.

**Review by albanD — APPROVED** [review]
Sounds pretty good to me.
We should cleanup public APIs before landing though.

File: docs/source/conf.py (hunk 2 above)
Comment (ezyang) [reply docs/source/conf.py:325]: It should get doc'ed. But TBH, I would prefer we call these SymInt in Python and not SymIntNode. (More comments from the peanut gallery? @Chillee @Krovatkin )

File: torch/csrc/jit/python/init.cpp (hunk 4 above)
Comment (ezyang) [reply torch/csrc/jit/python/init.cpp:165]: we have

```
using c10::SymFloat;
```

in the file.

File: torch/csrc/jit/python/pybind_utils.cpp (hunk 6 above)
Comment (ezyang) [reply torch/csrc/jit/python/pybind_utils.cpp:71]: The constant value doesn't matter, I just tried to make it something distinctive so that you'd notice if you were accidentally using it. If you like, we could also just zero initialize it, or max_int initialize it. Consistency unnecessary

File: docs/source/conf.py (hunk 2 above)
Comment (ezyang) [reply docs/source/conf.py:325]: I'd rather we do this in a separate PR, there's a bunch more doc stuff we'll have to do to make all of this make sense.

File: c10/core/SymFloat.h [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+#include <c10/util/Exception.h>
+#include <c10/util/intrusive_ptr.h>
+
+#include <limits>
+#include <memory>
+
+namespace c10 {
+
+// NB: this is actually double precision; we're using the Python naming here
+class C10_API SymFloat {
```
Comment (anjali411) [inline c10/core/SymFloat.h:14]: Why did you choose to use python naming here? It's ok with the comment but generally I would likely forget that this was meant to represent double if I saw this at an arbitrary line in the C++ code.

Comment (ezyang) [reply c10/core/SymFloat.h:14]: It's for consistency with SymInt, which is also Python naming (int is typically 32-bit) and the JIT type system which calls it float.
